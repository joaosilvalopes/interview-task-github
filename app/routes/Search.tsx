import { FormEvent, useEffect, useState } from 'react';
import { useNavigate, Outlet, useParams } from '@remix-run/react';

import { SearchContainer, Title, SearchForm, SearchInput, SearchButton } from './Search.styles';

type SearchFormOnSubmitEvent = FormEvent<HTMLFormElement> & {
  target: HTMLFormElement & {
    username: HTMLInputElement
  }
}

const Search = () => {
  const navigate = useNavigate();
  const { searchQuery } = useParams();
  const [loading, setLoading] = useState(false);

  const handleSubmit = (event: SearchFormOnSubmitEvent) => {
    event.preventDefault();

    const newSearchQuery = event.target.username.value;

    if(!newSearchQuery || searchQuery === newSearchQuery) return;

    setLoading(true);
    navigate(`/search/${newSearchQuery}`);
  };

  useEffect(() => {
    searchQuery && setLoading(false);
  }, [searchQuery]);

  return (
    <SearchContainer>
      <Title>Github User Search</Title>
      <SearchForm onSubmit={handleSubmit} data-testid="search-form">
        <SearchInput placeholder="username" type="text" id="username" name="username" defaultValue={searchQuery} data-testid="search-input" />
        <SearchButton type="submit" data-testid="search-button">{loading ? 'Loading' : 'Search'}</SearchButton>
      </SearchForm>
      <Outlet key={searchQuery} />
    </SearchContainer>
  );
};

export default Search;
