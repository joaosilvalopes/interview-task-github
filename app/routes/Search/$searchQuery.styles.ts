import styled from 'styled-components';
import { Link } from '@remix-run/react';

export const SearchResultList = styled.ul`
  list-style: none;
  margin: 2rem 0;
  padding: 0;
  width: 100%;
  max-width: 40rem;
`;

export const SearchResultListItem = styled.li`
  margin-bottom: 1.5rem;
  text-align: center;
`;

export const SearchResultLink = styled(Link)`
  display: block;
  padding: 1rem;
  text-decoration: none;
  font-size: 1.6rem;
  font-weight: bold;
  border-radius: 0.25rem;
  background-color: #eee;
  color: #333;
  transition: background-color 0.3s;

  &:hover {
    background-color: #ccc;
  }
`;
