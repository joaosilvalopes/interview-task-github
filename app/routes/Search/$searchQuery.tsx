import { useState, useEffect, useCallback } from 'react';
import { useLoaderData, useParams } from '@remix-run/react';
import { LoaderArgs } from '@remix-run/node';
import throttle from 'lodash.throttle';

import { SearchResultList, SearchResultListItem, SearchResultLink } from './$searchQuery.styles';

import searchUsers from '~/sdk/searchUsers';

import usePagination from '~/hooks/usePagination';

type SearchLoaderArgs = LoaderArgs & {
    params: {
        searchQuery: string;
    }
};

export const PAGE_SIZE = 100;

export const loader = async ({ params }: SearchLoaderArgs) => {
    const { searchQuery } = params;

    try {
        const { totalCount, usernames } = await searchUsers(searchQuery, 1, PAGE_SIZE);

        return { totalCount, usernames };
    } catch (e) {
        console.error(e);
    }
};

const SearchResults = () => {
    const { searchQuery } = useParams();
    const { totalCount, usernames: firstPageEntries } = useLoaderData();
    const [loadingUser, setLoadingUser] = useState<string>();

    const fetchPage = useCallback((page: number) => searchUsers(searchQuery as string, page, PAGE_SIZE).then((res) => res.usernames), [searchQuery]);

    const { 
        isPageLoading, 
        entries: usernames, 
        nextPage 
    } = usePagination<string>({
        fetchPage,
        firstPageEntries,
        pageSize: PAGE_SIZE,
        entryCount: totalCount,
    });

    useEffect(() => {
        const handleScroll = throttle(() => {
            const isAtPageBottom = window.innerHeight + document.documentElement.scrollTop === document.documentElement.offsetHeight;

            if (isAtPageBottom) {
                nextPage();
            }
        }, 100);

        document.addEventListener('scroll', handleScroll);

        return () => {
            document.removeEventListener('scroll', handleScroll);
        };
    }, []);

    return (
        <>
            {<p data-testid="result-count">Found {totalCount} results for {searchQuery}</p>}
            {usernames.length > 0 &&
                <SearchResultList>
                    {usernames.map((username) => (
                        <SearchResultListItem key={username}>
                            <SearchResultLink onClick={() => setLoadingUser(username)} to={`/user/${username}`} data-testid={`search-result-link-${username}`}>{loadingUser === username ? 'Loading' : username}</SearchResultLink>
                        </SearchResultListItem>
                    ))}
                </SearchResultList>}
            {isPageLoading && <p data-testid="loading-indicator">Loading...</p>}
        </>
    );
}

export default SearchResults;

