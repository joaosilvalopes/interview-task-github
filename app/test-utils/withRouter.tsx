import { useState, useEffect, ComponentType } from "react";
import { To, Router, Routes, Route } from "react-router";
import { History } from "@remix-run/router";

const withRouter = (Component: ComponentType, { history, path }: { history: History, path: string }) => () => {
    const [location, setLocation] = useState(history.location);

    useEffect(() => {
        history.listen(({ location }) => setLocation(location));
    }, []);

    const push = (to: To) => setTimeout(() => history.push(to), 100); // simulate remix loader fetch before changing route

    return <Router location={location} navigator={{ ...history, push }}>
        <Routes>
            <Route path={path} element={<Component />} />
            <Route path="*" element={null} />
        </Routes>
    </Router>;
};

export default withRouter;
