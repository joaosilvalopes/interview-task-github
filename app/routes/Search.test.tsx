import { createMemoryHistory, To } from '@remix-run/router';
import { render, fireEvent, waitFor } from '@testing-library/react';

import withRouter from '~/test-utils/withRouter';

import Search from './Search';

const history = createMemoryHistory({ initialEntries: ['/search'], v5Compat: true });

const SearchWithRouter = withRouter(Search, { path: '/search/:searchQuery?', history });

describe('Search component', () => {
  it('renders correctly and redirects to /search/:searchQuery on search form submit', async () => {
    const { getByTestId, asFragment } = render(<SearchWithRouter />);

    expect(getByTestId('search-form')).toBeInTheDocument();
    expect(getByTestId('search-input')).toBeInTheDocument();
    expect(getByTestId('search-button')).toBeInTheDocument();
    expect(getByTestId('search-button')).toHaveTextContent('Search');

    const searchQuery = 'testQuery';

    fireEvent.submit(getByTestId('search-form'), { target: { username: { value: searchQuery } } });

    expect(getByTestId('search-button')).toHaveTextContent('Loading');

    await waitFor(() => {
      expect(history.location.pathname).toBe(`/search/${searchQuery}`);
      expect(getByTestId('search-button')).toHaveTextContent('Search');
    });

    expect(asFragment()).toMatchSnapshot();
  });
});

