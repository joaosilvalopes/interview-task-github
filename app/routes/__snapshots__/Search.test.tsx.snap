// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Search component renders correctly and redirects to /search/:searchQuery on search form submit 1`] = `
<DocumentFragment>
  <main
    class="sc-beySbM ecnJdQ"
  >
    <h1
      class="sc-guDLRT hBSIrn"
    >
      Github User Search
    </h1>
    <form
      class="sc-dmyDGi lhIuuh"
      data-testid="search-form"
    >
      <input
        class="sc-hLQRIN dXWFfT"
        data-testid="search-input"
        id="username"
        name="username"
        placeholder="username"
        type="text"
        value="testQuery"
      />
      <button
        class="sc-eDLJxc jIgEkF"
        data-testid="search-button"
        type="submit"
      >
        Search
      </button>
    </form>
  </main>
</DocumentFragment>
`;
